prime.go

```diff
@@ -0,0 +1,198 @@
+package merkledag
+
+import (
+	"github.com/ipld/go-ipld-prime"
+	dagpb "github.com/ipld/go-ipld-prime-proto"
+)
+
+// Kind returns a value from the Kind enum describing what the
+// essential serializable kind of this node is (map, list, integer, etc).
+// Most other handling of a node requires first switching upon the kind.
+func (n *ProtoNode) Kind() ipld.Kind {
+	_, _ = n.EncodeProtobuf(false)
+	return n.encoded.Kind()
+}
+
+// LookupByString looks up a child object in this node and returns it.
+// The returned Node may be any of the Kind:
+// a primitive (string, int64, etc), a map, a list, or a link.
+//
+// If the Kind of this Node is not Kind_Map, a nil node and an error
+// will be returned.
+//
+// If the key does not exist, a nil node and an error will be returned.
+func (n *ProtoNode) LookupByString(key string) (ipld.Node, error) {
+	_, err := n.EncodeProtobuf(false)
+	if err != nil {
+		return nil, err
+	}
+	return n.encoded.LookupByString(key)
+}
+
+// LookupByNode is the equivalent of LookupByString, but takes a reified Node
+// as a parameter instead of a plain string.
+// This mechanism is useful if working with typed maps (if the key types
+// have constraints, and you already have a reified `schema.TypedNode` value,
+// using that value can save parsing and validation costs);
+// and may simply be convenient if you already have a Node value in hand.
+//
+// (When writing generic functions over Node, a good rule of thumb is:
+// when handling a map, check for `schema.TypedNode`, and in this case prefer
+// the LookupByNode(Node) method; otherwise, favor LookupByString; typically
+// implementations will have their fastest paths thusly.)
+func (n *ProtoNode) LookupByNode(key ipld.Node) (ipld.Node, error) {
+	_, err := n.EncodeProtobuf(false)
+	if err != nil {
+		return nil, err
+	}
+	return n.encoded.LookupByNode(key)
+}
+
+// LookupByIndex is the equivalent of LookupByString but for indexing into a list.
+// As with LookupByString, the returned Node may be any of the Kind:
+// a primitive (string, int64, etc), a map, a list, or a link.
+//
+// If the Kind of this Node is not Kind_List, a nil node and an error
+// will be returned.
+//
+// If idx is out of range, a nil node and an error will be returned.
+func (n *ProtoNode) LookupByIndex(idx int64) (ipld.Node, error) {
+	_, err := n.EncodeProtobuf(false)
```

Comment:
hopefully this is cached? is it preferable to have a `n.GetEncoded` that provides the pointer to the internal node and computes it if it doesn't exist?

Reply:
that's actually what the somewhat inaccurately named EncodeProtobuf actually does -- the boolean paramenter, assuming it's false, means don't re-encode unless the data has changed.